src/FromJSONTrait.php

```diff
@@ -27,9 +27,38 @@ public static function fromJSON($jsonStr) {
 
         if (is_null($cfg)) {
             $err = json_last_error();
-            throw new \InvalidArgumentException("Invalid JSON: $err");
+            throw new JsonParseErrorException($cfg, $err, self::provideJsonErrorMessageForCode($err));
         }
         $called_class = get_called_class();
         return new $called_class($cfg);
     }
+
+    private static function provideJsonErrorMessageForCode($code)
+    {
+
+        /**
+         * This is a hastily thrown together fix for the fact that
+         * php 5.4 doesn't have json_last_error_msg(). I very much recommend
+         * that this is replaced with json_last_error_msg(), is removed and the minimum
+         * recommended version of php is changed to 5.5. Especially seeing as php 5.4
+         * is now no longer being supported as of me writing this.
+         * @see http://php.net/supported-versions.php
+         */
```

Comment:
My main concern here is getting ahead of Moodle's require php version because it means that we have to require a higher PHP version for the Moodle plugins than for Moodle itself, and that's even assuming that the latest version of Moodle is being used. 

That said, the Moodle Logstore plugin already requires PHP 5.5 so perhaps that's not a good argument.